liberapay/models/participant.py

```diff
@@ -2329,6 +2332,15 @@ def update_goal(self, goal, cursor=None):
                  RETURNING receiving, npatrons
                 """, (self.id,))
                 self.set_attributes(**r._asdict())
+            if donations_paused or donations_resumed:
+                for tipper in tippers:
+                    tipper.send_email(
+                        'donations_paused',
+                        tipper.get_email(tipper.get_email_address()),
+                        recipient=self.username,
+                        donations_paused=donations_paused,
+                        donations_url=tipper.url('giving/')
+                    )
```

Comment:
Can this be accomplished by using Participant's notify() method? Also when I switched to using notify(), the request seemed to get stuck. Do you know why this might be happening?

also for the last point

> thirdly because the user should be able to reverse their decision before the emails are sent

If a decision is reversed, what should occur? Do the emails that were queued to send get dequeued?

Reply:
The `notify` method should be used, but in a separate function which would be either a payday method or a cron job, like `send_donation_reminder_notifications` or `send_account_disabled_notifications`.

Reply:
> Also when I switched to using notify(), the request seemed to get stuck. Do you know why this might be happening?

`update_goal` runs in a database transaction (which is confusingly called a "cursor" for legacy reasons), but `notify` doesn't, so if you tried to call `notify` from inside the `with` block you could have created a "deadlock" (two database workers trying to modify the same row, each one waiting for the other to complete its transaction before continuing).

> If a decision is reversed, what should occur? Do the emails that were queued to send get dequeued?

No. Once the email is queued for sending we don't cancel it. What we do instead is delay the creation of the notification which results in the email being queued for sending. Donors don't need to be notified immediately that donations have been paused, this kind of information can be delayed for an hour or even a week.